File: original-source/Sources/InputMonitor/Pipeline/MetricsAggregator.swift
import Foundation

/// 指标聚合器：在事件串行队列中被逐个喂入事件，维护当日增量与累计计数
/// - keyDown/keyUp 配对计算按住时长（剔除 ARepeat 重置）
/// - 事件间隔 ≤60s 视为连续活动，累加活动时间
/// - 跨天自动轮换计数器
final class MetricsAggregator {
    /// 连续活动允许的最大事件间隔（秒）
    var activeGapThresholdSeconds: TimeInterval = 60

    // MARK: - 当日累计（供 UI 实时展示）
    private(set) var day: String
    private(set) var keyCount = 0
    private(set) var clickCount = 0
    private(set) var scrollCount = 0
    private(set) var moveDistance = 0.0        // px
    private(set) var keyDurationMs: Int64 = 0
    private(set) var activeInputSeconds = 0.0
    private(set) var activeAppSeconds = 0.0

    // MARK: - 未落库增量
    private var deltaKeyCount = 0
    private var deltaClickCount = 0
    private var deltaScrollCount = 0
    private var deltaMoveDistance = 0.0
    private var deltaKeyDurationMs: Int64 = 0
    private var deltaActiveInputSeconds = 0.0
    private var deltaActiveAppSeconds = 0.0

    // MARK: - 配对与活动状态
    private var keyDownTimestamps: [Int64: UInt64] = [:]
    private var lastEventWallTime: Date?

    /// 互动分钟桶（当日）：任意来源活动（键鼠事件 ∪ 窗口心跳）按分钟去重
    private var activeMinutes = Set<Int>()

    init(day: String = EventRepository.dayString()) {
        self.day = day
    }

    /// 喂入一条事件（必须在同一串行队列调用）
    func process(_ record: InputEventRecord) {
        rotateIfNeeded(for: record.wallTime)
        accumulateActiveTime(at: record.wallTime)
        activeMinutes.insert(Self.minuteIndex(of: record.wallTime))

        switch record.kind {
        case .keyDown:
            guard !record.isAutoRepeat else { return }
            keyCount += 1
            deltaKeyCount += 1
            if let code = record.keyCode {
                keyDownTimestamps[code] = record.timestampNs
            }

        case .keyUp:
            guard let code = record.keyCode,
                  let downTs = keyDownTimestamps.removeValue(forKey: code) else { return }
            let durationNs = record.timestampNs &- downTs
            let ms = Int64(durationNs / 1_000_000)
            // 异常长按（>10min）多为系统休眠干扰，丢弃
            guard ms < 600_000 else { return }
            keyDurationMs += ms
            deltaKeyDurationMs += ms

        case .leftClick, .rightClick:
            clickCount += 1
            deltaClickCount += 1

        case .scroll:
            let lines = max(1, abs(record.scrollDelta))
            scrollCount += Int(lines)
            deltaScrollCount += Int(lines)

        case .mouseMoveSample:
            moveDistance += record.moveDelta
            deltaMoveDistance += record.moveDelta

        case .flagsChanged:
            break
        }
    }

    /// 应用心跳：累加窗口活动时间（heartbeatInterval 秒）；未锁屏时并入互动分钟桶
    func processAppHeartbeat(interval: TimeInterval) {
        rotateIfNeeded(for: Date())
        activeAppSeconds += interval
        deltaActiveAppSeconds += interval
        if !ScreenState.isLocked {
            activeMinutes.insert(Self.minuteIndex(of: Date()))
        }
    }

    /// 互动总秒数（分钟粒度，键鼠与窗口活动去重并集）
    var interactionSeconds: Double {
        Double(activeMinutes.count) * 60
    }

    /// 取走未落库增量（用于 mergeDailyStats）
    func drainDelta() -> (day: String, delta: DaySummary) {
        var delta = DaySummary(day: day)
        delta.keyCount = deltaKeyCount
        delta.clickCount = deltaClickCount
        delta.scrollCount = deltaScrollCount
        delta.keyDurationMs = deltaKeyDurationMs
        delta.moveDistance = deltaMoveDistance
        delta.activeInputSeconds = deltaActiveInputSeconds
        delta.activeAppSeconds = deltaActiveAppSeconds
        deltaKeyCount = 0
        deltaClickCount = 0
        deltaScrollCount = 0
        deltaKeyDurationMs = 0
        deltaMoveDistance = 0
        deltaActiveInputSeconds = 0
        deltaActiveAppSeconds = 0
        return (day, delta)
    }

    // MARK: - 快照（必须在事件串行队列上调用）

    struct Snapshot {
        let day: String
        let keyCount: Int
        let clickCount: Int
        let scrollCount: Int
        let moveDistance: Double
        let keyDurationMs: Int64
        let activeInputSeconds: Double
        let activeAppSeconds: Double
        let interactionSeconds: Double
    }

    func snapshot() -> Snapshot {
        Snapshot(
            day: day,
            keyCount: keyCount,
            clickCount: clickCount,
            scrollCount: scrollCount,
            moveDistance: moveDistance,
            keyDurationMs: keyDurationMs,
            activeInputSeconds: activeInputSeconds,
            activeAppSeconds: activeAppSeconds,
            interactionSeconds: interactionSeconds
        )
    }

    // MARK: - 私有

    /// 活动时间按墙上时间（wallTime）计算：对任何事件来源/时间基都一致可靠
    private func accumulateActiveTime(at wallTime: Date) {
        if let last = lastEventWallTime {
            let gap = wallTime.timeIntervalSince(last)
            if gap >= 0, gap <= activeGapThresholdSeconds {
                activeInputSeconds += gap
                deltaActiveInputSeconds += gap
            }
        }
        lastEventWallTime = wallTime
    }

    private func rotateIfNeeded(for date: Date) {
        let currentDay = EventRepository.dayString(for: date)
        guard currentDay != day else { return }
        // 跨天：重置计数器（增量由外部在跨天前 flush）
        day = currentDay
        keyCount = 0
        clickCount = 0
        scrollCount = 0
        moveDistance = 0
        keyDurationMs = 0
        activeInputSeconds = 0
        activeAppSeconds = 0
        keyDownTimestamps.removeAll()
        lastEventWallTime = nil
        activeMinutes.removeAll()
    }

    /// 当日分钟索引（0...1439）
    static func minuteIndex(of date: Date) -> Int {
        let comps = Calendar.current.dateComponents([.hour, .minute], from: date)
        return (comps.hour ?? 0) * 60 + (comps.minute ?? 0)
    }
}

File: current-integration/src/InputMonitor.Core/MetricsAggregator.cs
namespace InputMonitor.Core;

/// <summary>
/// In-memory daily counters. Must be fed from a single consumer thread.
/// </summary>
public sealed class MetricsAggregator
{
    public TimeSpan ActiveGapThreshold { get; set; } = TimeSpan.FromSeconds(60);

    public string Day { get; private set; }
    public int KeyCount { get; private set; }
    public int ClickCount { get; private set; }
    public int ScrollCount { get; private set; }
    public double MoveDistance { get; private set; }
    public long KeyDurationMs { get; private set; }
    public double ActiveInputSeconds { get; private set; }
    public double ActiveAppSeconds { get; private set; }

    private int _deltaKeyCount;
    private int _deltaClickCount;
    private int _deltaScrollCount;
    private double _deltaMoveDistance;
    private long _deltaKeyDurationMs;
    private double _deltaActiveInputSeconds;
    private double _deltaActiveAppSeconds;

    private readonly Dictionary<long, ulong> _keyDownTimestamps = [];
    private DateTimeOffset? _lastEventWallTime;
    private readonly HashSet<int> _activeMinutes = [];

    public MetricsAggregator(string? day = null)
    {
        Day = day ?? EventRepository.DayString(DateTimeOffset.Now);
    }

    public double InteractionSeconds => _activeMinutes.Count * 60d;

    public void Process(InputEventRecord record)
    {
        RotateIfNeeded(record.WallTime);
        AccumulateActiveTime(record.WallTime);
        _activeMinutes.Add(MinuteIndex(record.WallTime));

        switch (record.Kind)
        {
            case InputEventKind.KeyDown:
                if (record.IsAutoRepeat)
                {
                    return;
                }

                KeyCount++;
                _deltaKeyCount++;
                if (record.KeyCode is { } code)
                {
                    _keyDownTimestamps[code] = record.TimestampNs;
                }

                break;

            case InputEventKind.KeyUp:
                if (record.KeyCode is not { } upCode ||
                    !_keyDownTimestamps.Remove(upCode, out var downTs))
                {
                    return;
                }

                var ms = (long)((record.TimestampNs - downTs) / 1_000_000);
                if (ms is < 0 or >= 600_000)
                {
                    return;
                }

                KeyDurationMs += ms;
                _deltaKeyDurationMs += ms;
                break;

            case InputEventKind.LeftClick:
            case InputEventKind.RightClick:
                ClickCount++;
                _deltaClickCount++;
                break;

            case InputEventKind.Scroll:
                var lines = Math.Max(1, Math.Abs(record.ScrollDelta));
                ScrollCount += (int)lines;
                _deltaScrollCount += (int)lines;
                break;

            case InputEventKind.MouseMoveSample:
                MoveDistance += record.MoveDelta;
                _deltaMoveDistance += record.MoveDelta;
                break;
        }
    }

    public void ProcessAppHeartbeat(TimeSpan interval, DateTimeOffset now, bool screenLocked)
    {
        RotateIfNeeded(now);
        ActiveAppSeconds += interval.TotalSeconds;
        _deltaActiveAppSeconds += interval.TotalSeconds;
        if (!screenLocked)
        {
            _activeMinutes.Add(MinuteIndex(now));
        }
    }

    public void RestoreInteractionBaseline(IEnumerable<int> minutes)
    {
        foreach (var minute in minutes)
        {
            if (minute is >= 0 and <= 1439)
            {
                _activeMinutes.Add(minute);
            }
        }
    }

    public (string Day, DaySummary Delta) DrainDelta()
    {
        var delta = new DaySummary
        {
            Day = Day,
            KeyCount = _deltaKeyCount,
            ClickCount = _deltaClickCount,
            ScrollCount = _deltaScrollCount,
            KeyDurationMs = _deltaKeyDurationMs,
            MoveDistance = _deltaMoveDistance,
            ActiveInputSeconds = _deltaActiveInputSeconds,
            ActiveAppSeconds = _deltaActiveAppSeconds
        };
        _deltaKeyCount = 0;
        _deltaClickCount = 0;
        _deltaScrollCount = 0;
        _deltaKeyDurationMs = 0;
        _deltaMoveDistance = 0;
        _deltaActiveInputSeconds = 0;
        _deltaActiveAppSeconds = 0;
        return (Day, delta);
    }

    public MetricsSnapshot Snapshot()
    {
        return new MetricsSnapshot(
            Day,
            KeyCount,
            ClickCount,
            ScrollCount,
            MoveDistance,
            KeyDurationMs,
            ActiveInputSeconds,
            ActiveAppSeconds,
            InteractionSeconds);
    }

    private void AccumulateActiveTime(DateTimeOffset wallTime)
    {
        if (_lastEventWallTime is { } last)
        {
            var gap = (wallTime - last).TotalSeconds;
            if (gap is >= 0 and <= 60)
            {
                ActiveInputSeconds += gap;
                _deltaActiveInputSeconds += gap;
            }
        }

        _lastEventWallTime = wallTime;
    }

    private void RotateIfNeeded(DateTimeOffset date)
    {
        var currentDay = EventRepository.DayString(date);
        if (currentDay == Day)
        {
            return;
        }

        Day = currentDay;
        KeyCount = 0;
        ClickCount = 0;
        ScrollCount = 0;
        MoveDistance = 0;
        KeyDurationMs = 0;
        ActiveInputSeconds = 0;
        ActiveAppSeconds = 0;
        _keyDownTimestamps.Clear();
        _lastEventWallTime = null;
        _activeMinutes.Clear();
        _deltaKeyCount = 0;
        _deltaClickCount = 0;
        _deltaScrollCount = 0;
        _deltaKeyDurationMs = 0;
        _deltaMoveDistance = 0;
        _deltaActiveInputSeconds = 0;
        _deltaActiveAppSeconds = 0;
    }

    public static int MinuteIndex(DateTimeOffset date)
    {
        var local = date.ToLocalTime();
        return (local.Hour * 60) + local.Minute;
    }
}
